core/store/src/trie/mem/construction.rs

```diff
@@ -75,79 +78,137 @@ struct TrieConstructionSegment {
-                children[i as usize] = Some(child);
+            for (i, child) in self.children.iter() {
+                children[*i as usize] = Some(*child);
             }
-            if let Some(value) = self.value {
+            if let Some(value) = &self.value {
                 InputMemTrieNode::BranchWithValue { children, value }
             } else {
                 InputMemTrieNode::Branch { children }
             }
-        } else if let Some(value) = self.value {
+        } else if let Some(value) = &self.value {
             assert!(self.child.is_none());
             assert!(self.children.is_empty());
-            InputMemTrieNode::Leaf { value, extension: self.trail.into_boxed_slice() }
+            InputMemTrieNode::Leaf { value, extension: &self.trail }
         } else {
             assert!(self.child.is_some());
             assert!(self.children.is_empty());
-            InputMemTrieNode::Extension {
-                extension: self.trail.into_boxed_slice(),
-                child: self.child.unwrap(),
-            }
+            InputMemTrieNode::Extension { extension: &self.trail, child: self.child.unwrap() }
         };
         MemTrieNodeId::new(arena, input_node)
     }
 }
 
+/// A helper trait to make the construction code more readable.
+///
+/// Whenever we encode nibbles to vector, we want to use a vector from the
+/// freelist; otherwise allocation is quite slow.
+trait NibblesHelper {
+    fn encode_to_vec(&self, freelist: &mut VecU8Freelist, is_leaf: bool) -> ReusableVecU8;
+    fn encode_leftmost_to_vec(
+        &self,
+        freelist: &mut VecU8Freelist,
+        len: usize,
+        is_leaf: bool,
+    ) -> ReusableVecU8;
+}
+
+impl NibblesHelper for NibbleSlice<'_> {
+    fn encode_to_vec(&self, freelist: &mut VecU8Freelist, is_leaf: bool) -> ReusableVecU8 {
+        let mut vec = freelist.alloc();
+        self.encode_to(is_leaf, vec.vec_mut());
+        vec
+    }
+
+    fn encode_leftmost_to_vec(
+        &self,
+        freelist: &mut VecU8Freelist,
+        len: usize,
+        is_leaf: bool,
+    ) -> ReusableVecU8 {
+        let mut vec = freelist.alloc();
+        self.encode_leftmost_to(len, is_leaf, vec.vec_mut());
+        vec
+    }
+}
+
 impl<'a> TrieConstructor<'a> {
     pub fn new(arena: &'a mut Arena) -> Self {
-        Self { arena, segments: vec![] }
+        // We should only have as many allocations as the number of segments
+        // alive, which is at most the length of keys. We give a generous
+        // margin on top of that. If this is exceeded in production, an error
+        // is printed; if exceeded in debug, it panics.
+        const EXPECTED_FREELIST_MAX_ALLOCATIONS: usize = 4096;
+        Self {
+            arena,
+            segments: vec![],
+            trail_freelist: VecU8Freelist::new(EXPECTED_FREELIST_MAX_ALLOCATIONS),
+        }
+    }
+
+    fn recycle_segment(&mut self, segment: TrieConstructionSegment) {
```

Comment:
is this for optimization (freeing before the drop happens?)

Reply:
No, freeing is mandatory before dropping. Dropping without freeing will panic in debug.